File: packages/harbor-plugin/scripts/verify-candidate-docker.py
"""Internal driver for verify-candidate-docker.mjs; no provider credentials.

Runs the real Harbor DockerEnvironment and DshCandidateAgent lifecycle. The
Node parent owns the bounded, controlled production model Broker and authorizes
the first evaluation prompt only after checking that setup used zero requests.
"""
from __future__ import annotations

import argparse
import asyncio
import json
import shutil
import sys
import uuid
from pathlib import Path

from harbor.environments.docker.docker import DockerEnvironment
from harbor.models.agent.context import AgentContext
from harbor.models.task.config import EnvironmentConfig
from harbor.models.trial.paths import TrialPaths
from harbor_dsh_evolution.agent import DshCandidateAgent
from harbor_dsh_evolution.candidate import snapshot_candidate, verify_candidate


def report(stage: str, **fields: object) -> None:
    print(json.dumps({"stage": stage, **fields}), flush=True)


async def docker(*args: str) -> str:
    process = await asyncio.create_subprocess_exec(
        "docker", *args, stdout=asyncio.subprocess.PIPE, stderr=asyncio.subprocess.PIPE
    )
    stdout, stderr = await asyncio.wait_for(process.communicate(), timeout=30)
    if process.returncode:
        raise RuntimeError(f"Docker {args[0]} failed: {stderr.decode(errors='replace')}")
    return stdout.decode().strip()


async def verify(args: argparse.Namespace) -> None:
    root = Path(args.output).resolve(strict=True)
    source = Path(args.template).resolve(strict=True)
    candidate = root / "candidate"
    shutil.copytree(source, candidate, ignore=shutil.ignore_patterns("node_modules", ".harbor-runtime"))
    manifest = snapshot_candidate(candidate, candidate_id="harbor-keyless-docker-verification", version="1.0.0")
    assert manifest.runtime["transport"] == "acp"
    image_id = await docker("image", "inspect", args.image, "--format", "{{.Id}}")
    assert image_id.startswith("sha256:")
    session_id = "harbor-keyless-acp-" + uuid.uuid4().hex[:16]
    environment_dir = root / "environment"
    environment_dir.mkdir()
    paths = TrialPaths(root / "trial")
    paths.mkdir()
    paths.chmod_dir()
    environment = DockerEnvironment(
        environment_dir=environment_dir,
        environment_name=session_id,
        session_id=session_id,
        trial_paths=paths,
        task_env_config=EnvironmentConfig(docker_image=image_id, workdir="/workspace", cpus=1, memory_mb=2048),
        keep_containers=False,
        mounts=[
            {"type": "bind", "source": str(paths.agent_dir), "target": "/logs/agent"},
            {"type": "bind", "source": str(paths.verifier_dir), "target": "/logs/verifier"},
            {"type": "bind", "source": str(paths.artifacts_dir), "target": "/logs/artifacts"},
        ],
    )
    report("created", image_id=image_id, session_id=session_id, candidate_digest=manifest.digest)
    try:
        await asyncio.wait_for(environment.start(force_build=False), timeout=90)
        agent = DshCandidateAgent(
            logs_dir=paths.agent_dir,
            candidate_path=str(candidate),
            candidate_digest=manifest.digest,
            candidate_version=manifest.version,
            candidate_model_provider="controlled-fixture",
            candidate_model="controlled-model",
        )
        await asyncio.wait_for(agent.setup(environment), timeout=660)
        readiness = json.loads((paths.agent_dir / "candidate-runtime.json").read_text())
        assert readiness["candidate_digest"] == manifest.digest
        assert readiness["readiness"] == "initialize-and-session-new"
        assert readiness["preflight_model_requests"] == "disabled"
        report("setup_complete", readiness=readiness)
        permission = await asyncio.wait_for(asyncio.to_thread(sys.stdin.readline), timeout=15)
        if permission.strip() != "continue":
            raise RuntimeError("Parent did not authorize the controlled evaluation prompt")
        context = AgentContext()
        await asyncio.wait_for(
            agent.run("Reply exactly HARBOR_ACP_OK", environment, context), timeout=90
        )
        await environment.prepare_logs_for_host()
        agent.populate_context_post_run(context)
        assert context.metadata is not None
        identity = context.metadata["candidate"]
        assert identity["digest"] == manifest.digest
        assert identity["runtime"] == manifest.runtime
        assert context.metadata["acp"]["prompt_response"]["stopReason"] == "end_turn"
        trajectory_path = paths.agent_dir / "trajectory.json"
        trajectory = json.loads(trajectory_path.read_text())
        assert "HARBOR_ACP_OK" in json.dumps(trajectory)
        assert trajectory.get("steps"), "Real ACP events must produce an ATIF trajectory"
        launcher = (paths.agent_dir / "acp-launch.sh").read_text()
        assert "npx" not in launcher and "@latest" not in launcher
        assert f"/{manifest.runtime['entrypoint']}" in launcher
        assert context.metadata["acp"]["selected_distribution"] == "candidate-local"
        await environment.download_dir("/opt/harbor-dsh-candidate/.sessions", root / "candidate-sessions")
        events = [
            json.loads(line)
            for path in (root / "candidate-sessions").rglob("*.jsonl")
            for line in path.read_text().splitlines()
            if line.strip()
        ]
        assert any(event.get("type") == "assistant/message" and "HARBOR_ACP_OK" in json.dumps(event) for event in events)
        assert any(event.get("type") == "turn/end" and event["data"]["reason"]["kind"] == "completed" for event in events)
        verify_candidate(candidate, expected_digest=manifest.digest)
        report("run_complete", candidate=identity, trajectory_file=str(trajectory_path),
               atif_steps=len(trajectory["steps"]), persisted_events=len(events),
               selected_distribution="candidate-local", prompt_stop_reason="end_turn")
    finally:
        # delete=False still removes this exact Compose project's containers and
        # network, but never requests image deletion or touches another project.
        await asyncio.wait_for(environment.stop(delete=False), timeout=60)
        containers = await docker("ps", "-aq", "--filter", f"label=com.docker.compose.project={session_id}")
        networks = await docker("network", "ls", "-q", "--filter", f"label=com.docker.compose.project={session_id}")
        assert not containers, f"Owned verification containers remain for {session_id}"
        assert not networks, f"Owned verification networks remain for {session_id}"
        assert await docker("image", "inspect", image_id, "--format", "{{.Id}}") == image_id
        report("cleanup_complete", session_id=session_id, containers_remaining=0, networks_remaining=0, image_preserved=True)


if __name__ == "__main__":
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--image", required=True)
    parser.add_argument("--template", required=True)
    parser.add_argument("--output", required=True)
    asyncio.run(verify(parser.parse_args()))

File: packages/harbor-plugin/scripts/verify-candidate-docker.mjs
#!/usr/bin/env node
/** Real Python adapter + Linux Docker + npm-lock closure, controlled model only. */
import assert from 'node:assert/strict'
import { execFile, spawn } from 'node:child_process'
import { mkdtemp, mkdir, writeFile } from 'node:fs/promises'
import { tmpdir } from 'node:os'
import { dirname, join, resolve } from 'node:path'
import { fileURLToPath } from 'node:url'
import { parseArgs, promisify } from 'node:util'
import { CandidateModelRuntime } from '../../dsh-plugin/lib/model-runtime.js'

const runFile = promisify(execFile)
const here = dirname(fileURLToPath(import.meta.url))
const pluginRoot = resolve(here, '..')
const { values } = parseArgs({ options: {
  image: { type: 'string' },
  python: { type: 'string', default: join(pluginRoot, '.venv/bin/python') },
  template: { type: 'string', default: join(pluginRoot, 'src/harbor_dsh_evolution/runtime_template') },
} })
assert(values.image, 'Pass --image <already-built-local-image>; this script never builds or publishes images')
const output = await mkdtemp(join(tmpdir(), 'harbor-real-docker-acp-'))
await mkdir(join(output, 'docker-config'))
const { stdout: endpoint } = await runFile('docker', ['context', 'inspect', '--format', '{{.Endpoints.docker.Host}}'])
assert(endpoint.trim().startsWith('unix://'), 'This verification only supports an explicit local Docker Unix socket')
const { stdout: pluginInfo } = await runFile('docker', ['info', '--format', '{{json .ClientInfo.Plugins}}'])
const compose = JSON.parse(pluginInfo).find(plugin => plugin.Name === 'compose' && !plugin.Err)
assert(compose?.Path, 'Docker Compose must already be installed')
// Retain only the installed CLI-plugin location, not registry auth or other
// Host Docker configuration. The Unix socket is supplied explicitly below.
await writeFile(join(output, 'docker-config/config.json'), JSON.stringify({ cliPluginsExtraDirs: [dirname(compose.Path)] }))
const requests = []
let setupComplete = false
let runComplete = false
let cleanupComplete = false
let driver
let lease
let driverExited
const milestones = []
const runtime = new CandidateModelRuntime({
  agentDefaultModel: { currentSelection: () => ({ provider: 'controlled-fixture', model: 'controlled-model' }) },
  llm: {
    listProviders: () => [{ id: 'controlled-fixture', name: 'Controlled fixture; no vendor' }],
    resolveModelInfo: async () => ({ id: 'controlled-model', name: 'Controlled model', context: { contextWindow: 32768 } }),
    async * stream(options) {
      assert(setupComplete, 'The real adapter readiness check must not call a model')
      requests.push(options)
      assert.equal(requests.length, 1, 'Only one controlled evaluation request is authorized')
      assert(JSON.stringify(options).includes('Reply exactly HARBOR_ACP_OK'))
      assert(JSON.stringify(options).includes('evidence-driven research assistant'), 'Candidate persona must survive the production overlay')
      yield { type: 'block-start', index: 0, blockType: 'text' }
      yield { type: 'text-delta', index: 0, text: 'HARBOR_ACP_OK' }
      yield { type: 'block-end', index: 0, block: { type: 'text', text: 'HARBOR_ACP_OK' } }
      yield { type: 'finish', reason: { kind: 'stop' } }
    },
  }, get: () => undefined,
}, { modelBrokerBindHost: '0.0.0.0', modelBrokerAdvertisedHost: 'host.docker.internal', modelBrokerMaxRequests: 1, modelBrokerMaxRequestBytes: 1048576 })

try {
  lease = await runtime.openLease(await runtime.resolve(), { candidateDigest: 'controlled-docker-acceptance', jobName: 'keyless-real-docker-acp', maxRequests: 1 })
  const environment = {
    PATH: dirname(process.execPath) + ':/usr/local/bin:/usr/bin:/bin:/opt/homebrew/bin',
    LANG: 'C.UTF-8', DOCKER_HOST: endpoint.trim(), DOCKER_CONFIG: join(output, 'docker-config'),
    PYTHONPATH: join(pluginRoot, 'src'), PYTHONUNBUFFERED: '1',
    HSE_MODEL_GATEWAY_URL: lease.endpoint, HSE_MODEL_GATEWAY_TOKEN: lease.token,
    HSE_MODEL_GATEWAY_PROVIDER: lease.candidateProvider,
    HSE_MODEL_GATEWAY_PROTOCOL: lease.protocol,
    HSE_MODEL_GATEWAY_INFO: JSON.stringify(lease.modelInfo),
  }
  driver = spawn(resolve(values.python), [join(here, 'verify-candidate-docker.py'),
    '--image', values.image, '--template', resolve(values.template), '--output', output,
  ], { env: environment, stdio: ['pipe', 'pipe', 'pipe'] })
  let stderr = ''
  let pending = ''
  let protocolError
  driver.stderr.setEncoding('utf8').on('data', text => { stderr += text })
  driver.stdout.setEncoding('utf8').on('data', text => {
    pending += text
    const lines = pending.split('\n')
    pending = lines.pop()
    for (const line of lines.filter(Boolean)) {
      try {
        const milestone = JSON.parse(line)
        milestones.push(milestone)
        if (milestone.stage === 'setup_complete') {
          assert.equal(requests.length, 0)
          assert.equal(lease.usage().modelRequests, 0)
          setupComplete = true
          driver.stdin.write('continue\n')
        }
        if (milestone.stage === 'run_complete') runComplete = true
        if (milestone.stage === 'cleanup_complete') cleanupComplete = true
        console.log(JSON.stringify(milestone))
      } catch (error) {
        protocolError = error
        driver.stdin.end('stop\n')
      }
    }
  })
  driverExited = new Promise((resolveExit, reject) => {
    driver.once('error', reject)
    driver.once('close', (code, signal) => resolveExit({ code, signal }))
  })
  const exit = await driverExited
  driver = undefined
  await writeFile(join(output, 'driver-stderr.log'), stderr)
  if (protocolError) throw protocolError
  assert.equal(exit.code, 0, `Python adapter verification failed (${JSON.stringify(exit)}):\n${stderr}`)
  assert(setupComplete && runComplete && cleanupComplete, 'All actual lifecycle and cleanup gates must complete')
  assert.equal(requests.length, 1)
  assert.equal(lease.usage().modelRequests, 1)
  const report = { status: 'passed', evidence: 'real-python-adapter-linux-docker-locked-npm-controlled-model',
    setup_model_requests: 0, evaluation_model_requests: 1, real_model_provider_calls: 0,
    output, milestones,
  }
  await writeFile(join(output, 'verification.json'), JSON.stringify(report, null, 2) + '\n')
  console.log(JSON.stringify(report, null, 2))
} finally {
  if (driver) {
    driver.stdin.end('stop\n')
    // The Python driver owns its container's finally cleanup. Do not replace
    // that with a broad docker prune or a signal that skips its cleanup.
    if (driverExited) await driverExited
  }
  await lease?.close()
  console.error(`Docker acceptance evidence: ${output}`)
}
